proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,139 @@
+- Start Date: 2022-10-25
+- Reference Issues: TODO
+- Implementation PR: <!-- leave empty -->
+
+# Summary
+
+Currently, Astro's [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) enables developers to switch between `'static'` (default) and `'server'` outputs. One of our most requested features is to enable route-level control of `output` so that some routes can be server-rendered while others are statically generated.
+
+# Example
+
+This RFC proposes a new, opt-in `output` mode named `'hybrid'`.
+
+```js
+import { defineConfig } from 'astro/config'
+import node from '@astrojs/node'
+
+export default defineConfig({
+  output: 'hybrid',
+  adapter: node()
+})
+```
+
+For individual routes, `'static'` will be the default. To opt-in to `'server'` output, users may `export const output: 'static' | 'server'`.
+
+```astro
+---
+export const output = 'server'
+
+const text = Astro.url.searchParams.get('text') ?? 'Hello world!';
+---
+
+<h1 set:html={text} />
+```
+
+# Motivation
+
+Since Astro v1.0.0 was released, we've gotten consistent feedback that more granular control over `output` is a requirement for scaling projects. Users would like to statically render some routes and server render others, reducing the need for full static rebuilds for frequently changing data.
+
+One very common use-case is a fully static site that also exposes `api/` endpoints which are handled on the server. Another common example is a largely static site that can generate a few dynamic pages at request time (for example, a user profile page).
+
+The eventual solution should enable granular control over `output` on a per-route basis. Since Astro defaults to `'static'` output, this should remain the default in the eventual solution as well.
+
+# Detailed design
+
+## Public API
+
+There are two public API surfaces to this RFC:
+
+- Updating the `astro.config.mjs` to accept a new `output` value, `'hybrid'`. The existing values of `'static' | 'server'` would remain supported. A value of `'hybrid'` will enabled detection for the following feature:
```

Comment:
I'm still not totally sold on whether we need a build-level solution here or if cache headers will cover most use cases

Cache headers are really tricky to deal with, especially when if I'm living in an SSG world most days.  I could see a nice first step here being a nicer API for setting cache headers that adds better DX with hints on the different cache header options

```js
Astro.cookies.cache({ duration: '1 year', public: true, httpOnly: true })`
```

Not sure exactly what the final API would need to be, but in this idea our type definitions could help explain when the `public` or `httpOnly` flags should be used.  We kept the `"1 year"` shorthand out of the `Astro.cookies` API to start with, this might be a good time to pull that in 🤔 

**Pros**

- no build or adapters changes needed
- `static` and `server` work as-is today
- we could add a proper hybrid mode as a follow-up if caching doesn't cut it for enough users, the `Astro.cookies.cache` helper would still be handy

**Cons**
- still has the first request performance hit, this dodges the true hybrid use case
- improving the DX of cache headers may be a lost cause...

Reply:
> I'm still not totally sold on whether we need a build-level solution here or if cache headers will cover most use cases

My use-case for a build-level solution is to pre-build certain resources and then omit the dependencies necessary to build those resources from the production server. For instance, I generate some social images by rendering them using Puppeteer, but don't want to ship that dependency or the large browser it requires to prod. Having a way to generate those routes at build-time and then omit unused dependencies would be ideal.

(this applies to large dependencies, but also dependencies that are Node-only if I'm deploying to a non-Node server runtime)

Reply:
Thanks @jacobwgillespie! I think that use case makes a very compelling argument for _not_ doing this solely through cache headers, though we definitely have discussed that as another API that would be generally useful.